resources/puppetlabs/puppetdb/pql/pql-grammar-experimental.ebnf

```diff
@@ -0,0 +1,139 @@
+(* PQL experimental EBNF specification
+
+   This file is the formal specification used by our parsing library that defines
+   the PQL grammar for translation to the AST query language.
+
+   This file is used by Mark Engelberg's 'Instaparse' parser in clojure, and
+   follows its own form of EBNF, so is not strictly 'EBNF' but an augmentation. *)
+
+(* TODO: the following items are just not supported in the syntax yet:
+
+   - group_by
+   - aggregate functions
+   - ~>
+   - in <array> *)
+
+(* TODO: the following features are awaiting PDB AST changes first, but
+   generally have support within PDB in one form or another:
+
+   - order by
+   - limit, offset
+   - include_total *)
+
+(* TODO: the follow items need fundamental changes to PDB first:
+
+   - descending into hashes, like facts & resource parameters conveniently
+   - modifying child data returned (ie. document aware projection) *)
+
+(* Top level query, this is the entry point for queries *)
+<query> = from;
+
+(* From - Setting context and providing filtering and projection *)
+from = entity, [<whitespace>], where, [[<whitespace>], extract];
```

Comment:
I was considering moving the field list for projection to the front, like so:

```
nodes[certname] { certname = 'foo' }
```

As opposed to:

```
nodes { certname = 'foo'} [certname]
```

Just feels better after typing some examples over and over, and keeps the column names next to the entity type instead of way at the end after the filter.

Any opinions?